feat: Make the as prop accept all kind of UrlLike or string

martpie opened the pull request:
Should fix #517

```tsx
<Link
  href='/somepage'
  as={{ pathname: '/test', hash: 'something', query: { page: 2 } }}
>
  Test
</Link>
```

will output now something like: 

```
http://localhost:3000/en/test?page=2#something
```

martpie commented:
I have an additional question: would you accept a PR to enable `strict: true` in TypeScript? I will happily work on making it work, I already saw a couple of potential bugs covered by `strictNullChecks` etc

isaachinman commented:
> enable strict: true in TypeScript?

Yes.

martpie commented:
I'll do it in a separate pr then :)

martpie commented:
Do you need anything else from me @isaachinman ?

isaachinman commented on __tests__/utils/lng-path-corrector.test.ts:
```diff
@@ -35,7 +35,7 @@ describe('lngPathCorrector utility function', () => {
   it('throws if currentRoute.as is not undefined or a string', () => {
```
Needs to be updated.

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -1,26 +1,34 @@
   if (asType === 'undefined') {
     as = formatUrl(href, { unicode: true })
-  } else if (asType === 'string') {
+  } else if (asType === 'string' || (asType === 'object' && originalAs !== null)) {
```
Is checking the type against `object` safe here? Many, many things are `objects` in JavaScript.

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -1,26 +1,34 @@
 }
 
-const parseHref = (originalHref) => {
+const parseHref = (originalHref): UrlObject => {
```
Is this typing correct? Can't it also be a string?

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
   Object.values(localeSubpaths).forEach((subpath) => {
-    if (subpathIsPresent(as, subpath)) {
-      as = removeSubpath(as, subpath)
+    if (typeof as === 'object') {
```
Shouldn't we also check that `as.pathname` is a string?

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
     const subpath = subpathFromLng(config, currentLanguage)
 
-    as = `/${subpath}${currentAs}`.replace(/\/$/, '')
+    if (typeof as === 'object') {
```
Same feedback here for `as.pathname`.

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
+
     href.query.lng = currentLanguage
-    href.query.subpath = subpath
+    href.query.subpath = subpath || ''
```
Can you explain why this is necessary?

isaachinman commented on src/utils/lng-path-corrector.ts:
```diff
@@ -58,19 +66,37 @@ export default (config: Config, currentRoute, currentLanguage) => {
+    }
+
+    if (typeof href.query !== 'object') throw new Error(`'href.query' must be an object`)
+    if (!href.query) href.query = {}
```
Besides the formatting, which needs to be fixed, this looks pretty dangerous to me. A falsy check isn't good enough - let's be clear about our assertions here.

Also, you're throwing if `href.query` isn't an `object` the line before this. Can you explain the logic here?

martpie commented:
Sorry for the delay, I will try to address some of your comments and reply to some others asap

> These changes make me a little nervous as they increase complexity and seem to be relying on some unclear/unsafe type assertions.

Regarding the complexity, maybe I can try to simplify things a little bit and be a bit less imperative/sequential.

Regarding the unsafe type assertions, you're right and I will try to improve these. I based myself a lot on TypeScript type checking + Next.js's typings, but there may be runtime issues indeed.

isaachinman commented:
Hi @martpie - do you want to proceed with this one?

martpie commented:
By proceeding, do you mean taking it over? or discarding it? 😄

isaachinman commented:
@martpie Sure, I can take it over if you don't want to continue. Just let me know.

martpie commented:
Dropping it, it's a bit outdated and I will restart a branch from scratch, addressing some PR comments. I will also check how `next` implemented their link component as there may be some interesting things to get there.
